staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/client.go

```diff
@@ -0,0 +1,212 @@
+	vaultConfig := vaultapi.DefaultConfig()
+	vaultConfig.Address = config.Address
+
+	tlsConfig := &vaultapi.TLSConfig{
+		CACert:        config.CACert,
+		ClientCert:    config.ClientCert,
+		ClientKey:     config.ClientKey,
+		TLSServerName: config.TLSServerName,
+	}
+	if err := vaultConfig.ConfigureTLS(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return vaultapi.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to vaultapi.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig) error {
+	switch {
+	case config.ClientCert != "" && config.ClientKey != "":
+		token, err := c.tlsToken(config)
+		if err != nil {
+			return err
+		}
+		c.client.SetToken(token)
+	case config.RoleID != "":
+		token, err := c.appRoleToken(config)
+		if err != nil {
+			return err
+		}
+		c.client.SetToken(token)
+	default:
+		// configuration has already been validated, flow should not reach here
+		return errors.New("the Vault authentication configuration is invalid")
+	}
+
+	return nil
+}
+
+func (c *clientWrapper) tlsToken(config *EnvelopeConfig) (string, error) {
+	resp, err := c.client.Logical().Write("/"+path.Join(c.authPath, "cert", "login"), nil)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) appRoleToken(config *EnvelopeConfig) (string, error) {
+	data := map[string]interface{}{
+		"role_id":   config.RoleID,
+		"secret_id": config.SecretID,
+	}
+	resp, err := c.client.Logical().Write("/"+path.Join(c.authPath, "approle", "login"), data)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) decrypt(keyName string, cipher string) (string, error) {
+	var result string
+
+	data := map[string]string{"ciphertext": cipher}
+	resp, err := c.request(path.Join(c.decryptPath, keyName), data)
+	if err != nil {
+		return result, err
+	}
+
+	result, ok := resp.Data["plaintext"].(string)
+	if !ok {
+		return result, fmt.Errorf("failed type assertion of vault decrypt response type: %v to string", reflect.TypeOf(resp.Data["plaintext"]))
+	}
+
+	return result, nil
+}
+
+func (c *clientWrapper) encrypt(keyName string, plain string) (string, error) {
+	var result string
+
+	data := map[string]string{"plaintext": plain}
+	resp, err := c.request(path.Join(c.encryptPath, keyName), data)
+	if err != nil {
+		return result, err
+	}
+
+	result, ok := resp.Data["ciphertext"].(string)
+	if !ok {
+		return result, fmt.Errorf("failed type assertion of vault encrypt response type: %v to string", reflect.TypeOf(resp.Data["ciphertext"]))
+	}
+
+	return result, nil
+}
+
+// This request check the response status code. If get code 403, it sets forbidden true.
+func (c *clientWrapper) request(path string, data interface{}) (*vaultapi.Secret, error) {
+	req := c.client.NewRequest("POST", "/"+path)
+	if err := req.SetJSONBody(data); err != nil {
+		return nil, err
+	}
+
+	resp, err := c.client.RawRequest(req)
+	if resp != nil {
```

Comment:
nit: check the error first

Reply:
The API used returns both error and resp for this case, so response is checked first.

Reply:
When an Go API returns an error you always check the error to determine if the other returned values are valid. If `err != nil` and `resp != nil` I would expect `resp` to be invalid unless the API specifically stated otherwise.

Reply:
The API, does not document it explicitly. So I changed the code to check error first. The error returned here (by vault client API) does not have a specific interface. So I am parsing error message to get the HTTP status code.